[ML][PYTORCH] Upgrade to LibTorch 1.9 and use the new inference mode

From the PyTorch [release notes](https://github.com/pytorch/pytorch/releases)

> New APIs to optimize performance and packaging for model inference deployment

Described here inference mode is a new RAII guard optimised for inference workloads: 
https://github.com/ailzhang/rfcs/blob/rfc0011/RFC-0011-InferenceMode.md

Building the Pytorch 1.9 in the Linux Aarch64 with `USE_MKLDNN=ON` and `BLAS=Eigen` failed with an internal compiler error. 

```
[ 77%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/ReduceAllOpsKernel.cpp.DEFAULT.cpp.o
during RTL pass: reload
In file included from /usr/src/pytorch/aten/src/ATen/ATen.h:13,
                 from /usr/src/pytorch/aten/src/ATen/native/UnaryOps.h:3,
                 from /usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:1:
/usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp: In function 'at::native::DEFAULT::lgamma_kernel(at::TensorIteratorBase&)::<lambda()>::<lambda()>::<lambda(char**, const int64_t*, int64_t)>':
/usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:636:11: internal compiler error: Max. number of generated reload insns per insn is achieved (90)

  636 |           }
      |           ^
/usr/src/pytorch/aten/src/ATen/Dispatch.h:66:12: note: in definition of macro 'AT_PRIVATE_CASE_TYPE_USING_HINT'
   66 |     return __VA_ARGS__();                                                        \
      |            ^~~~~~~~~~~
/usr/src/pytorch/aten/src/ATen/Dispatch.h:242:7: note: in expansion of macro 'AT_PRIVATE_CASE_TYPE'
  242 |       AT_PRIVATE_CASE_TYPE(NAME, at::ScalarType::Float, float, __VA_ARGS__)    \
      |       ^~~~~~~~~~~~~~~~~~~~
/usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:642:5: note: in expansion of macro 'AT_DISPATCH_FLOATING_TYPES_AND'
  642 |     AT_DISPATCH_FLOATING_TYPES_AND(kBFloat16, iter.dtype(), #op "_vml_cpu", [&]() { \
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:644:11: note: in expansion of macro 'IMPLEMENT_ITERATOR_LAMBDA'
  644 |           IMPLEMENT_ITERATOR_LAMBDA(op),                                            \
      |           ^~~~~~~~~~~~~~~~~~~~~~~~~
/usr/src/pytorch/build/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp:799:1: note: in expansion of macro 'IMPLEMENT_FLOAT_KERNEL'
  799 | IMPLEMENT_FLOAT_KERNEL(lgamma)
      | ^~~~~~~~~~~~~~~~~~~~~~
0xa13207 lra_constraints(bool)
	../../gcc-9.3.0/gcc/lra-constraints.c:4901
0x9fd877 lra(_IO_FILE*)
	../../gcc-9.3.0/gcc/lra.c:2472
0x9bcc67 do_reload
	../../gcc-9.3.0/gcc/ira.c:5523
0x9bcc67 execute
	../../gcc-9.3.0/gcc/ira.c:5707
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
gmake[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp.DEFAULT.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2
gmake: *** [all] Error 2
Building wheel torch-1.9.0
```

--- a/bin/pytorch_inference/Main.cc
+++ b/bin/pytorch_inference/Main.cc
@@ -63,7 +63,7 @@ torch::Tensor infer(torch::jit::script::Module& module,
                                              inputSize, at::dtype(torch::kInt64)));
     }
 
-    torch::NoGradGuard noGrad;
+    torch::InferenceMode inferenceModeGuard;
     auto result = module.forward(inputs);
     if (result.isTuple()) {
         // For BERT models the result tensor is the first element in a tuple

